### PR TITLE
Added functionality to efficiently acquire and release service instances

### LINKED ISSUE #797: Design of ServiceUtil is broken
I just wanted to check out the new `ServiceUtil` class and found out that it is completely broken:

* It doesn't compile anymore on Java 8

```java
final ServiceReference<?>[] refs = bundleContext.getServiceReferences(clazz.getName(), filter);
@SuppressWarnings("unchecked")
final ServiceReference<T>[] reference = (refs == null ? new ServiceReference[0] : refs);
```

OSGi has provides proper methods for returning `Collection<ServiceReference<T>>` which should be used instead.

* The call for `get` and `unget` is completely imbalenced:

A call to `getServiceReferences` returns all service **references** while a call to `ungetServiceReferences` actually **un**gets the services. However many calls to `getServiceReferences` ignore most of the returned services but do return other instances.

So assume you have the following services and their respective counters:

* Start
A=0, B=0, C=0

* call to get -> eval, return B -> unget
A=0, B=0, C=-1
Still B is returned and actively references in the code. Although its counter is zero now.

* call to get -> unget
A=-1, B=-1, C=-2

Even worse, assume that C actually did have a reference. Then the call to get -> unget would have actually reset that reference to zero.

### PR DESCRIPTION
**Pull request by amitjoy**
This PR comprises :

1. `AssetService` and `DriverService` API enhancements
2. Code Refactorings
3. New functionality to acquire and release service instances in an easy-to-use way
4. Removed NO-OP methods from Wire Service implementation

Also fixes #797

### CODE REVIEW

**Comment by amitjoy** [comment]
@markoer Could you kindly check once why the `testCreateWires(..)` test case in this PR is failing? It seems to me that it should have passed.

File: kura/org.eclipse.kura.util/src/main/java/org/eclipse/kura/util/service/ServiceSupplier.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        return new ServiceSupplier<>(reference);
+    }
+
+    /**
+     * Retrieves the list of target service instances
+     *
+     * @return the list of service instances or empty list
+     */
+    @Override
+    public List<T> get() {
```
Comment (ctron) [inline kura/org.eclipse.kura.util/src/main/java/org/eclipse/kura/util/service/ServiceSupplier.java]: Calling the get method multiple times will cause the counter to be incremented multiple times. However the close method will only release once.

So again this implementation seems unbalanced to me.

Comment (amitjoy) [reply kura/org.eclipse.kura.util/src/main/java/org/eclipse/kura/util/service/ServiceSupplier.java]: You are right. I have just added a check to validate if the service instance has already been retrieved from the service reference.

**Comment by amitjoy** [comment]
Any update?

**Comment by cdealti** [comment]
@ctron can you please review? You debated a lot about this in #797.
@amitjoy can you please take a look to the table created by @nicolatimeus at the end of #253?
Just to be all on the same page...

**Comment by MMaiero** [comment]
Closing, for now, as it is too far behind develop.
We will recover it, if needed

**Pull request closed without merging**